docs/30_configuration.md
# Proxy Manager

This service proxies HTTP calls to external services.

## Environment variables

The service use the following environment variables:

- **CONFIGURATION_PATH** (required for *static* configuration): the file path of the service configuration file;
- **CONFIGURATION_FILE_NAME** (required for *static* configuration): the filename of the service configuration file (without the extension);
- **CONFIGURATION_URL** (required for *dynamic* configuration): the url of the CRUD collection with the service configuration;
- **PROXY_CACHE_TTL** (optional, default to `0`): the time to live (in seconds) of cached proxy configurations. This is used only with *dynamic* configuration.
- **LOG_LEVEL** (optional, default to `info`): level of the log. It could be trace, debug, info, warn, error, fatal;
- **HTTP_PORT** (optional, default to `8080`): port where the web server is exposed;
- **SERVICE_PREFIX** (optional): path prefix for all the specified endpoints (different from the status routes);
- **EXPOSE_MANAGEMENT_APIS** (optional, default `false`): allows to control whether or not [management APIs](./20_how_to_use.md#management-api) are exposed by the service (please note that this flag can be used only when running in dynamic configuration mode);
- **ALLOW_PROXY_OPTIMIZER** (optional, default to `true`): boolean that enables optimized proxy using reverse proxy and preventing saving body request in memory. Be careful, this optimization does not perform any retry, thus it is strongly suggested to configure the token validation endpoint in your proxy configuration;
- **DELAY_SHUTDOWN_SECONDS** (optional, default to `10` seconds): seconds to wait before starting the graceful shutdown. This delay is required in k8s to await for the DNS rotation;
- **DISABLE_PROXY_CACHE** (optional, default `false`): allows to disable the in-memory proxy cache. Such cache is used to prevent excessive CRUD invocations (useful when you want to prevent any possible out-dated cache hit whenever the service is using Dynamic Configuration in the context of a scaled HA architecture)
- **BASE_PATH_MATCHERS** (optional): comma-separated list of paths that are used to let the service handle complex base paths when using dynamic proxy configuration (note: this function is not compatible with the non-optimized proxy configuration, make sure to set `ALLOW_PROXY_OPTIMIZER=true`).
- **ADDITIONAL_HEADERS_TO_REDACT** (optional): comma separated values of additional headers to redact when logging. The following headers are always redacted: `Authorization`, `Cookie`, `Proxy-Authorization`, `Set-Cookie` and `Www-Authenticate`
- **HEADER_BLOCK_LIST** (optional): comma separated list of header names useful to prevent them from being forwarded to the target service. These headers will prevail if same headers are specified on other Environment Variables
- **HEADERS_TO_REMAP** (optional): comma separated list of `key:value` pairs where `key` is the Key of a header in the incoming request and `value` is the Key of the header in the proxied request.
- **DISALLOWED_RESPONSE_CONTENT_TYPE_LIST** (optional): comma separated list of values for the `Content-Type` header that the are disallowed in the response of the target service
- **ALLOWED_PROXY_TARGET_URLS** (optional): comma separated list of base URLs. Only proxies with targetBaseUrl specified in this list can be created and handled. If none is specified all target URLs are allowed.
- **TOKEN_PREEMPTIVE_EXPIRY_SECONDS** (optional, default to `30`): number of seconds before the actual token expiration to preemptively treat the token as expired and request a new one. This prevents race conditions where a token expires between the moment it's validated and when it's actually used in the HTTP request. For example, with a 30-second preemptive expiry, a token expiring at 12:00:30 will be considered expired at 12:00:00, forcing a token refresh and avoiding 401 errors.

:::caution
**ALLOW_PROXY_OPTIMIZER** will be dismissed with the next major release. The not optimized proxy functionality and the retry feature is deprecated and will be dismissed too.
:::

## Static configuration

This service requires a configuration file that provides all the different details regarding the external services to be proxied.
The configuration file can be mounted into the service either as a *ConfigMap* or as a *Secret* (for more details, please refer to this [documentation](../../development_suite/api-console/api-design/services#custom-configuration)).

:::caution
In case the former method (*Config Map*) is selected, please use [variables interpolation](../../development_suite/api-console/api-design/services#environment-variable-configuration) for sensitive data, such as:

- username
- password
- clientId
- clientSecret

This prevents to store those sensitive values as plain text in the project repository.
:::

### Configuration schema

The configuration must follow this schema:

```json
{
  "type": "object",
  "properties": {
    "proxies": {
      "type": "array",
      "items": {
        "type": "object",
        "required": ["targetBaseUrl","basePath"],
        "properties": {
          "authentication": {
            "type": "string",
            "enum": ["none","oauth2"],
            "default": "none"
          },
          "username": {
            "type": "string"
          },
          "password": {
            "type": "string"
          },
          "clientId": {
            "type": "string"
          },
          "clientSecret": {
            "type": "string"
          },
          "tokenIssuerUrl": {
            "type": "string"
          },
          "tokenIssuerValidationPath": {
            "type": "string"
          },
          "targetBaseUrl": {
            "type": "string",
            "pattern": "^https?:\\/\\/[a-zA-Z0-9.:_-]+(\\/((\\{[a-zA-Z0-9_-]+\\})|[a-zA-Z0-9_-]+))*\\/?$"
          },
          "basePath": {
            "type": "string",
            "pattern": "^\\/[a-zA-Z0-9_-]+(\\/((\\{[a-zA-Z0-9_-]+\\})|[a-zA-Z0-9_-]+))*\\/?$"
          },
          "grantType": {
            "type": "string",
            "enum": ["client_credentials","password"],
            "default": "client_credentials"
          },
          "authType": {
            "type": "string",
            "enum": ["client_secret_basic"],
            "default": "client_secret_basic"
          },
          "additionalAuthFields": {
            "type": "object",
            "default": null
          },
          "headersToProxy": {
            "type": "array",
            "items": {
              "type": "string"
            },
            "default": []
          },
          "additionalHeaders": {
            "type": "array",
            "items": {
              "type": "object",
              "properties": {
                "name": {
                  "type": "string"
                },
                "value": {
                  "type": "string"
                }
              },
              "required": ["name","value"]
            },
            "default": []
          }
        }
      }
    }
  }
}
```

The **proxies** array contains one item for each external service that has to by proxied.

A proxy can have the following fields:

- **targetBaseUrl**: the url of the external service. This is a required field and has to start with an *http* or *https* scheme. Possible path parameters from the base path can be referenced and added to the url with the `{param-name}` syntax (**only for static configuration**).
- **basePath**: the name of the related endpoint exposed by the *Proxy Manager*. This is a required field and has to start with a `/`.
Path parameter can be specified with the `{param-name}` syntax and eventually forwarded to the external service in the *targetBaseUrl* using the same name (**only for static configuration**).
- **authentication**: the type of authentication done by the proxy, which can either be `oauth2` or `none`.
- **username**: the user identifier for the OAuth2 authentication (only [Password Grant](https://oauth.net/2/grant-types/password/) flow).
- **password**: the user password used in case of OAuth2 authentication (only [Password Grant](https://oauth.net/2/grant-types/password/) flow).
- **clientId**: the client identifier used in case of OAuth2 authentication.
- **clientSecret**: the client secret used in case of OAuth2 authentication.
- **tokenIssuerUrl**: the authorization server url that has to be called to obtain an access token.
- **tokenIssuerValidationUrl**: the authorization server url that has to be called to validate an access token. This validation is performed at each API call. If not provided, the validation is skipped and the proxy only checks for token expiration.
- **grantType**: the type of procedure used to retrieve an access token. At the moment, the service supports the following OAuth2 Grant Types:
  - [*client_credentials*](https://oauth.net/2/grant-types/client-credentials/)
  - [*password*](https://oauth.net/2/grant-types/password/)
- **authType**: the method used to stuff the client credentials in the request when asking for a new access token. This is required only for the Client Credential Grant flow. At the moment, the service only supports the *client_secret_basic* type.
- **additionalAuthFields**: an object containing additional fields to be used in the authentication request. These fields are added in the request body performed to acquire the access token. Both object keys and values must be of type string.
- **headersToProxy**: a list of headers that must be forwarded when calling the external service. The default behavior, triggered when this field is not provided, is to forward all the headers of original request. In case the list is empty, no header from original request is proxied.
- **additionalHeaders**: a list of headers the must be added to the request when calling the external service, after authentication is successful.
A use case of additionalHeaders is api keys management (e.g. x-api-key header).

:::caution
Path parameters inside **`targetBaseUrl`** and **`basePath`** are only allowed for the static configuration.
:::

## Dynamic configuration

The service requires a CRUD collection (named as you prefer) that provides all the different details regarding the external services to be proxied: each document **must** match the *proxy* schema specified in the [configuration schema](#configuration-schema).

:::tip
By default, when using the *dynamic configuration* only the first segment of the invoked path is used as base path to extract configurations from the database; in this scenario path parameters are not supported.

**E.g.**: without `BASE_PATH_MATCHERS`, given a request with path `/one/two/three`, the **`basePath`** searched on CRUD is `/one`.

Use the `BASE_PATH_MATCHERS` environment variable to provide one ore more base path matchers and let the service be able to extract more complex
base paths from the database.

Assuming you set `BASE_PATH_MATCHERS=/my-path/:param,/another-path/test-api` and then invoke `/my-path/123/some-other-api`, the **`basePath`** searched on CRUD will be `/my-path/123`.
:::

In order to configure correctly the CRUD collection, you can **import** the fields from this <a download target="_blank" href="docs_files_to_download/crud.fields.json">file</a>. This file already enables the Client Side Field Level Encryption (CSFLE) for those fields with sensitive data.

## Recommendations

It is recommended to add the following indexes to your CRUD collection:

- a *unique* index for field **`basePath`** (to guarantee proxy uniqueness),
- a search index for fields **`basePath`** and **`__STATE__`** (to improve performances).

## Configuration example

In this example, the *Proxy Manager* is configured to proxy requests to three external services.

- The first one is located at `external-service.com`, requires OAuth2 authentication (*client_credentials* grant type) and can be reached through the proxy with a call to the `/external-service` endpoint.
- The second one is located at `mia-client-credentials.com`, requires OAuth2 authentication (*client_credentials* grant type) and can be reached through the proxy with a call to the `/mia-service` endpoint. It employs additional fields that are added in the request to retrieve an access token.
- The third one is located at `legacy-service.com`, requires OAuth2 authentication (*password* grant type) and can be reached through the proxy with a call to the `/legacy-service` endpoint.
- The forth service is located at `other-service.com/{id}`, requires no authentication and can be reached with a call to the `/other-service/{id}` endpoint. For example, with a call to the `/other-service/page-id/additional-path` endpoint will be called the target service `other-service.com/page-id/additional-path`. Path parameters can be used only with static configuration.
- The latter service is located at `another-service.com`, requires no authentication, can be reached with a call to the `/another-service` endpoint, only selected headers are forwarded to it and the additional header `x-api-key` with `header-value` is sent.

```json
{
  "proxies": [
    {
      "authentication": "oauth2",
      "clientId": "6779ef20e75817b79602",
      "clientSecret": "GBAyfVL7YWtP6gudLIjbRZV_N0dW",
      "tokenIssuerUrl": "http://external-service.com/auth/oauth/token",
      "targetBaseUrl": "https://external-service.com",
      "basePath": "/external-service",
      "grantType": "client_credentials"
    },
    {
      "authentication": "oauth2",
      "clientId": "6739ef20e75817a79c02",
      "clientSecret": "GBAweVL7YWtP6gudLIjbRZV_NdW",
      "tokenIssuerUrl": "http://mia-client-credentials.com/auth/oauth/token",
      "tokenIssuerValidationPath": "/token-info",
      "targetBaseUrl": "https://mia-service.com",
      "basePath": "/mia-service",
      "grantType": "client_credentials",
      "additionalAuthFields": {
        "audience": "mia-audience"
      }
    },
    {
      "authentication": "oauth2",
      "username": "username",
      "password": "password",
      "clientId": "0fbd65b60ca0388f2069",
      "clientSecret": "ZUogSDuesC6VpZM9P2_mPTX4",
      "tokenIssuerUrl": "http://legacy-service.com/auth/oauth/token",
      "targetBaseUrl": "http://legacy-service.com",
      "basePath": "/legacy-service",
      "grantType": "password"
    },
    {
      "targetBaseUrl": "http://other-service.com/{id}",
      "basePath": "/other-service/{id}"
    },
    {
      "targetBaseUrl": "http://another-service.com",
      "basePath": "/another-service",
      "headersToProxy": [
        "Accept",
        "Content-Type"
      ],
      "additionalHeaders": [
        {
          "name": "x-api-key", 
          "value": "header-value"
        }
      ]
    }
  ]
}
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## 3.6.0 - 30-07-2025

### Added

- added `TOKEN_PREEMPTIVE_EXPIRY_SECONDS` environment variable to configure preemptive token expiry buffer, preventing race conditions between token validation and usage (default: 30 seconds)

### Fixed

- fixed OAuth2 token race conditions that could cause 401 errors when tokens expire between validation check and actual request

## 3.5.0 - 2025-01-23

### Fixed

- exposed management routes under `SERVICE_PREFIX`

### Added

- added `DISALLOWED_RESPONSE_CONTENT_TYPE_LIST` env to prevent specified `Content-Type` headers in the response of the target service.
- added `ALLOWED_PROXY_TARGET_URLS` env to specify which target URLs are allowed as targetBaseUrl for a proxy.

## 3.4.4 - 13-01-2025

### Updated

- `GET /-/proxies` management API supports filtering for multiple `basePath`. This can be either a string or a comma-separated list of strings

## 3.4.3 - 2024-12-18

### Fixed

- fixed optimized mode that was ignored on `"authentication": "none"`
- ensuring not following redirect on optimized mode

## 3.4.2 - 2024-12-16

### Fixed

- patch api failure if proxy authentication is set to invalid value
- return `tokenIssuerUrl` and `tokenIssuerValidationUrl` from GET management API

## 3.4.1 - 11-12-2024

### Updated

- better patch request payload validation

## 3.4.0 - 09-12-2024

### Added

- added `HEADERS_TO_REMAP` env to remap specified headers into other specified header key.
- added `HEADER_BLOCK_LIST` env to prevent specified headers from being forwarded to the target URL.
- ci: sbom generation

## 3.3.3 - 2024-12-02

- internal context management for crud dependency

## 3.3.2 - 29-11-2024

### Fixed

- fixed API responses preventing empty strings to be returned if fields are missing

## 3.3.1 - 27-11-2024

### Fixed

- `DISABLE_PROXY_CACHE`: correctly prevents cache to be written when disabled

## 3.3.0 - 22-11-2024

### Added

- `BASE_PATH_MATCHERS`: to make it possible to fetch complex base paths when using dynamic configuration rather than taking the first segment only.
- `DISABLE_PROXY_CACHE`: to make it possible to disable the in-memory proxy cache when using dynamic configuration.

## 3.2.0 - 15-11-2024

### Added

- Added GET `/-/proxies` API
- Added POST `/-/proxies` API
- Added DELETE `/-/proxies` API
- Added PATCH `/-/proxies/:proxyId` API
- Added PATCH `/-/proxies` API

### Refactor

- internal component refactor and improved standard library usage

### Fixed

- sanitize target base url trailing slash to prevent `//` token in final path

## 3.1.0 - 2024-10-29

### Added

- `ADDITIONAL_HEADERS_TO_REDACT` environment variable to set additional headers to redact in logs

### Security

- Redacted sensitive headers in logs

## 3.0.0 - 2023-09-14

### BREAKING CHANGES

- remove trailing slash to handle requests without path parameters. Using the following configuration and calling the path `/docs` the proxy-manager will call `https://docs.mia-platform.eu/docs` and no more `https://docs.mia-platform.eu/docs/`: 
```javascript
{
  "targetBaseUrl": "https://docs.mia-platform.eu/docs",
  "basePath": "/docs"
}
```

### Added

- add request body log


## 2.0.0 - 2023-05-24

### BREAKING CHANGES

- `SERVICE_PREFIX` must match the following regex `^/[a-zA-Z0-9_-]+$`
- remove entire `basePath` from request path and not only the first segment. Using the following configuration and calling the path `/mia/docs/fast-data` the proxy-manager will call `https://docs.mia-platform.eu/fast-data` and no more `https://docs.mia-platform.eu/docs/fast-data`:
```javascript
{
  "targetBaseUrl": "https://docs.mia-platform.eu/",
  "basePath": "/mia/docs"
}
```

## 1.6.2 - 2023-05-23

### Fixed

- remove configuration `targetBaseUrl` wrong update

## 1.6.1 - 2023-05-19

### Fixed

- `retrieveUrl` correctly interpolates path parameters in case of `SERVICE_PREFIX`

## 1.6.0 - 2023-05-18

### Added

- add `additionalHeaders` to configuration to handle headers that will be proxied to target url

### Changed

- `targetBaseUrl` and `basePath` can contain path parameters

## 1.5.1 - 29-09-2022

### Changed

- update AccessToken.ExpiresIn type from int to json.RawMessage

## 1.5.0 - 1/09/2022

### Added

- Support to [dynamic configuration](https://git.tools.mia-platform.eu/platform/core/proxy-manager/-/issues/4): proxies are fetched from a CRUD collection 

## 1.4.1 - 31-07-2022

### Fixed

- [JMRBA](https://makeitapp.atlassian.net/browse/JMRBA-139): the service now is able to proxy the query parameters

### Changed

- delete token from cache when response returns an HTTP 403 error

## 1.4.0 - 29-04-2022

### Added

- logic to refresh a token if its ExpiresAt date is set and the ExpiresAt date is earlier than current moment

### Fixed

- added a _modify response_ logic to reverse proxy, so that when an HTTP 401 error is encountered
 cached access token is removed. This allow to clean up token cache upon token expiration
 when parameter `TokenIssuerValidationUrl` is not set

### Changed

- updated service dependencies

## 1.3.0 - 25-11-2021

### Changed

- debug logs added to improve the troubleshooting experience
- [RJMR-36](https://makeitapp.atlassian.net/browse/RJMR-36): the service now support optimized proxy without saving request body in memory, this feature is controlled by `ALLOW_PROXY_OPTIMIZER` environment variables and does not perform any retry upon request failures
- [JMRBA-84](https://makeitapp.atlassian.net/browse/JMRBA-84): fix proxy manager optimized with rewrite req host and x-forwarded-for header

## 1.2.0 - 18-10-2021

### Added

- introduced a mechanism that allows, for each proxy config, to select whether all the headers

internal/config/config.go
/**
 * Copyright 2026 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package config

import (
	_ "embed"
	"fmt"
	"regexp"
	"strings"

	"proxy-manager/entities"

	envlib "github.com/caarlos0/env/v11"
	"github.com/mia-platform/configlib"
)

//go:embed config.schema.json
var configSchema string

// EnvironmentVariables struct with the mapping of desired environment variables.
type EnvironmentVariables struct {
	LogLevel                       string            `env:"LOG_LEVEL" envDefault:"info"`
	HTTPPort                       string            `env:"HTTP_PORT" envDefault:"8080"`
	ServicePrefix                  string            `env:"SERVICE_PREFIX"`
	ServiceVersion                 string            `env:"SERVICE_VERSION"`
	ServiceConfigPath              string            `env:"CONFIGURATION_PATH"`
	ServiceConfigFileName          string            `env:"CONFIGURATION_FILE_NAME"`
	ServiceConfigUrl               string            `env:"CONFIGURATION_URL"`
	ProxyCacheTTL                  int               `env:"PROXY_CACHE_TTL" envDefault:"0"`
	AllowProxyOptimizer            bool              `env:"ALLOW_PROXY_OPTIMIZER" envDefault:"true"`
	DelayShutdownSeconds           int               `env:"DELAY_SHUTDOWN_SECONDS" envDefault:"10"`
	AdditionalHeadersToRedact      []string          `env:"ADDITIONAL_HEADERS_TO_REDACT"`
	HeaderBlockList                []string          `env:"HEADER_BLOCK_LIST"`
	HeadersToRemap                 map[string]string `env:"HEADERS_TO_REMAP" envSeparator:"," envKeyValSeparator:":"`
	ExposeManagementAPIs           bool              `env:"EXPOSE_MANAGEMENT_APIS" envDefault:"false"`
	DisableProxyCache              bool              `env:"DISABLE_PROXY_CACHE" envDefault:"false"`
	BasePathExtractorPrefixes      []string          `env:"BASE_PATH_MATCHERS"`
	DisallowedResponseContentTypes []string          `env:"DISALLOWED_RESPONSE_CONTENT_TYPE_LIST"`
	AllowedProxyTargetURLs         []string          `env:"ALLOWED_PROXY_TARGET_URLS" envSeparator:","`
	TokenPreemptiveExpirySeconds   int               `env:"TOKEN_PREEMPTIVE_EXPIRY_SECONDS" envDefault:"30"`
}

func GetEnvVariables() (EnvironmentVariables, error) {
	env, err := envlib.ParseAs[EnvironmentVariables]()
	if err != nil {
		return env, err
	}

	if (env.ServiceConfigPath == "" || env.ServiceConfigFileName == "") && env.ServiceConfigUrl == "" {
		return env, fmt.Errorf("required env variables not set")
	}
	if env.ServiceConfigPath != "" && env.ServiceConfigFileName != "" && env.ServiceConfigUrl != "" {
		return env, fmt.Errorf("cannot enable both dynamic and static configuration")
	}
	if len(env.ServicePrefix) > 0 {
		re := regexp.MustCompile(`^\/[a-zA-Z0-9_-]+$`)
		if !re.MatchString(env.ServicePrefix) {
			return env, fmt.Errorf("service prefix does not match the following regex: ^/[a-zA-Z0-9_-]+$")
		}
	}
	return env, nil
}

func (e EnvironmentVariables) IsStaticConfiguration() bool {
	return e.ServiceConfigPath != "" && e.ServiceConfigFileName != ""
}

func (e EnvironmentVariables) IsDynamicConfiguration() bool {
	return e.ServiceConfigUrl != ""
}

type ServiceConfig struct {
	Proxies []*entities.Proxy `json:"proxies" koanf:"proxies"`
}

func GetServiceConfigSchema() []byte {
	return []byte(configSchema)
}

func LoadServiceConfiguration(path, fileName string) (*ServiceConfig, error) {
	var config ServiceConfig
	err := configlib.GetConfigFromFile(fileName, path, GetServiceConfigSchema(), &config)
	if err != nil {
		return nil, err
	}
	for _, proxyConfiguration := range config.Proxies {
		pathParamsConfigIsCorrect, err := targetUrlAndBasePathContainSamePathParameters(proxyConfiguration)
		if !pathParamsConfigIsCorrect {
			return nil, err
		}
	}
	return &config, err
}

func targetUrlAndBasePathContainSamePathParameters(proxyConfiguration *entities.Proxy) (bool, error) {
	re := regexp.MustCompile(`{[a-zA-Z0-9_-]+}`)
	pathParameters := re.FindAllString(proxyConfiguration.TargetBaseUrl, -1)
	for _, pathParameter := range pathParameters {
		if !strings.Contains(proxyConfiguration.BasePath, pathParameter) {
			return false, fmt.Errorf("missing path parameters inside %v", proxyConfiguration.BasePath)
		}
	}
	return true, nil
}
